### PR TITLE
feat(codegen): save output to file if specified

### PR DESCRIPTION
**Pull request by mxschmitt**

### CODE REVIEW

File: src/cli.ts [hunk 1]
```diff
@@ -245,14 +246,13 @@ async function openPage(context: playwright.BrowserContext, url: string | undefi
...
-async function open(options: Options, url: string | undefined, enableRecorder: boolean) {
+async function open(options: Options, url: string | undefined, enableRecorder: boolean, outputFile?: string) {
   const { context, browserName, launchOptions, contextOptions } = await launchContext(options, false);
-  new ScriptController(browserName, launchOptions, contextOptions, context, process.stdout, enableRecorder, options.device);
+  new ScriptController(browserName, launchOptions, contextOptions, context, process.stdout, enableRecorder, outputFile, options.device);
+  await openPage(context, url);
   if (process.env.PWCLI_EXIT_FOR_TEST) {
-    await context.close()
-    process.exit(0)
+    await Promise.all(context.pages().map(p => p.close()))
```
Comment (pavelfeldman) [inline src/cli.ts:254]: super-nit: trim those {} around single-liners

File: src/cli.ts [hunk 2]
```diff
@@ -290,8 +290,8 @@ async function pdf(options: Options, captureOptions: CaptureOptions, url: string
   await browser.close();
 }
 
-async function codegen(options: Options, url: string | undefined) {
-  return open(options, url, true);
+async function codegen(options: Options, url: string | undefined, outputFile: string | undefined) {
```
Comment (pavelfeldman) [inline src/cli.ts:293]: nit: outputFile?: string

File: src/recorderController.ts [hunk 3]
```diff
@@ -50,6 +50,7 @@ export class RecorderController {
       for (const timer of this._timers)
         clearTimeout(timer);
       this._timers.clear();
+      this._output.writeOutputFile();
```
Comment (pavelfeldman) [inline src/recorderController.ts:53]: would typically call this `output.flush`

File: src/terminalOutput.ts [hunk 4]
```diff
@@ -121,14 +137,16 @@ export class TerminalOutput {
     if (eraseLastAction && this._lastActionText)
       eraseLines += this._lastActionText.split('\n').length;
     // And we erase the last action too if augmenting.
-    for (let i = 0; i < eraseLines; ++i)
+    for (let i = 0; i < eraseLines; ++i) {
       this._out.write('\u001B[1A\u001B[2K');
+      this._toBeWrittenLines.pop()
```
Comment (pavelfeldman) [inline src/terminalOutput.ts:142]: `pop();` you Python fan :P

File: src/terminalOutput.ts [hunk 5]
```diff
@@ -48,7 +53,18 @@ export class TerminalOutput {
...
+
+  writeOutputFile() {
+    if (this._outputFile)
+      fs.writeFileSync(this._outputFile, this._toBeWrittenLines.join('\n'))
+  }
+
+  private _write(text: string, suffix: string = "") {
+    if (this._outputFile)
+      this._toBeWrittenLines.push(...(text + suffix).trimEnd().split('\n'))
+    this._out.write(this._highlight(text) + suffix)
```
Comment (pavelfeldman) [inline src/terminalOutput.ts:67]: Let's throw a bit more design here: have 'fileOutput' and 'terminalOutput' generator 'Observers'. Observer will have { addLine:, popLine:, flush: } interface. Highlight will be implementation detail of the terminalOutput, with the highlight library only imported from there.

Comment (mxschmitt) [reply src/terminalOutput.ts:67]: good catch.

File: src/output.ts [hunk 6]
```diff
@@ -14,14 +14,13 @@
...
-import * as querystring from 'querystring';
 import * as playwright from 'playwright';
 import { Writable } from 'stream';
+import { Frame } from 'playwright';
 import { quote, Formatter } from './formatter';
 import { Action, actionTitle, NavigationSignal, PopupSignal, Signal, DownloadSignal, DialogSignal } from './recorderActions';
 import { MouseClickOptions, toModifiers } from './utils';
-import { highlight } from 'highlight.js';
-import { Frame } from 'playwright';
+import {OutputMultiplexer, TerminalOutput, FileOutput} from './outputs'
```
Comment (pavelfeldman) [inline src/output.ts:23]: supper-nit: space after { before }

File: src/outputs.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import * as querystring from 'querystring';
+import { Writable } from 'stream';
+import { highlight } from 'highlight.js';
+
+export class OutputMultiplexer {
+  private _outputs: IOutput[]
+  constructor() {
+    this._outputs = []
+  }
+  add(output: IOutput) {
```
Comment (pavelfeldman) [inline src/outputs.ts:11]: nit: Give us some blank lines between methods

Comment (pavelfeldman) [inline src/outputs.ts:11]: nit: pass array in the constructor - adding one half-way does not make much sense.

**Review by pavelfeldman — COMMENTED** [review]
This looks so much better!

File: src/output.ts [hunk 8]
```diff
@@ -30,14 +29,17 @@ export type ActionInContext = {
...
   private _lastAction: ActionInContext | undefined;
   private _lastActionText: string | undefined;
-  private _out: Writable;
+  private _outputs: OutputMultiplexer;
 
-  constructor(browserName: string, launchOptions: playwright.LaunchOptions, contextOptions: playwright.BrowserContextOptions, out: Writable, deviceName: string | undefined) {
-    this._out = out;
+  constructor(browserName: string, launchOptions: playwright.LaunchOptions, contextOptions: playwright.BrowserContextOptions, out: Writable, outputFile: string | undefined, deviceName: string | undefined) {
+    this._outputs = new OutputMultiplexer()
+    this._outputs.add(new TerminalOutput(out))
```
Comment (pavelfeldman) [inline src/output.ts:40]: Steps to make it better design-wise, based on the principle of least knowledge:
- Rename 'Output' to 'Generator'
- Declare 'GeneratorOutput' beside it, as far as Generator is concerned it emits into the GeneratorOutput.
- Instantiate Multiplexer externally to the Generator. You can already see how `out` and `outputFile` are only used to create it. Generator should not know about multiplexing.
- Use `this._output`, single, it is just an output sink

File: test/codgen.spec.ts [hunk 9]
```diff
@@ -82,4 +86,28 @@ it('should print the correct context options when using a device and additional
 })();`;
   await cli.waitFor(expectedResult);
   expect(cli.text()).toContain(expectedResult);
+});
+
+it('should save the codegen output to a file if specified', async ({ runCLI, tmpDir }) => {
```
[the remark this answers is not in the document]
Comment (mxschmitt) [reply test/codgen.spec.ts:91]: Won't work since it does contain a dynamic variable: `${emptyHTML}`

[the remark this answers is not in the document]
Comment (pavelfeldman) [reply test/codgen.spec.ts:91]: You sanitize it (replace it with EMPTY_URL). Not for this patch though.

**Review by pavelfeldman — APPROVED** [review]

**Comment by sjurgis** [comment]
Unsure this works, see comment in https://github.com/microsoft/playwright-cli/issues/74#issuecomment-696479157

**Pull request merged**